packages/vendure-scripts/src/assign-these-prodcuts-to-channel.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  AssetService,
+  ChannelService,
+  FacetService,
+  ID,
+  Injector,
+  Product,
+  ProductVariantService,
+  RequestContext,
+  SearchService,
+  TransactionalConnection,
+} from '@vendure/core';
+import { getSuperadminContextInChannel } from '../../test/src/admin-utils';
+
+export async function assignTheseProductsToChannel(
+  targetChannelId: ID,
+  injector: Injector,
+  products: Product[],
+  ctx: RequestContext
+): Promise<void> {
+  // start the transaction
+  const conn = injector.get(TransactionalConnection);
+  await conn.startTransaction(ctx);
+  // assign them to target channel
+  const productVariantService = injector.get(ProductVariantService);
+  const assetService = injector.get(AssetService);
+  const facetService = injector.get(FacetService);
+  const channelService = injector.get(ChannelService);
+  const searchService = injector.get(SearchService);
+  const facetIds = products
+    .map((product) =>
+      (product.facetValues ?? []).map((facetValue) => facetValue.facet.id)
+    )
+    .flat();
+  const assetIds = products
+    .map((product) => (product.assets ?? []).map((asset) => asset.id))
+    .flat();
+  const variantIds = products
+    .map((product) => (product.variants ?? []).map((variant) => variant.id))
+    .flat();
+  const targetChannel = await channelService.findOne(ctx, targetChannelId);
+  const superadminContextInTargetChannel = await getSuperadminContextInChannel(
+    injector,
+    targetChannel!
+  );
+  await Promise.all([
+    //we will first assign the product variants, which will also assign the associated products and variants' assets
+    productVariantService.assignProductVariantsToChannel(ctx, {
+      channelId: targetChannelId,
+      productVariantIds: variantIds,
+    }),
+    //we will then assign the product assets to the target channels
+    assetService.assignToChannel(ctx, { assetIds, channelId: targetChannelId }),
+    //we will then assign the facets with the associated values
+    facetService.assignFacetsToChannel(ctx, {
+      channelId: targetChannelId,
+      facetIds,
+    }),
+    searchService.reindex(superadminContextInTargetChannel),
```

Comment:
All these calls need to be awaited, even though they are inside a Promise.all. When you don't await them, they sort of dissapear in the background, and thus will not be awaited by the promise.all